lib/tree.rb

```diff
@@ -2,59 +2,129 @@ class TreeNode
   attr_reader :key, :value
   attr_accessor :left, :right
 
-   def initialize(key, val)
+  def initialize(key, val)
     @key = key
     @value = val
     @left = nil
     @right = nil
-   end
+  end
 end
 
 class Tree
   attr_reader :root
+
   def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: Olog(n)
+  # Space Complexity: O(1)
   def add(key, value)
-    raise NotImplementedError
+    new_node = TreeNode.new(key, value)
+
+    if @root == nil
+      @root = new_node
+    else
+      current_node = @root
+      previouse_node = @root
+
+      while current_node != nil
+        previouse_node = current_node
+        if key < current_node.key
+          current_node = current_node.left
+        else
+          current_node = current_node.right
+        end
+
+        if key < previouse_node.key
+          previouse_node.left = new_node
+        else
+          previouse_node.right = new_node
+        end
+      end
+
+      return new_node
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
   def find(key)
-    raise NotImplementedError
+    def find_helper(key, current_node)
```

Comment:
Making an inner method like this in Ruby is discouraged and messy.  It leads to `tree.rb:54: warning: method redefined; discarding old find_helper`